sweepbatcher/greedy_batch_selection.go

```diff
@@ -0,0 +1,290 @@
+package sweepbatcher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/txscript"
+	sweeppkg "github.com/lightninglabs/loop/sweep"
+	"github.com/lightningnetwork/lnd/input"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwallet/chainfee"
+)
+
+// greedyAddSweep selects a batch for the sweep using the greedy algorithm,
+// which minimizes costs, and adds the sweep to the batch.
+func (b *Batcher) greedyAddSweep(ctx context.Context, sweep *sweep) error {
+	if b.customFeeRate == nil {
+		return errors.New("greedy batch selection algorithm requires " +
+			"setting custom fee rate provider")
+	}
+
+	// Collect weight and fee rate info about the sweep and new batch.
+	sweepEntity, newBatchEntity, err := estimateSweepWeight(sweep)
+	if err != nil {
+		return fmt.Errorf("failed to estimate tx weight for "+
+			"sweep %x: %w", sweep.swapHash[:6], err)
+	}
+
+	// Collect weight and fee rate info about existing batches.
+	batches := make([]entity, 0, len(b.batches))
+	for _, batch := range b.batches {
+		batchEntity, err := estimateBatchWeight(batch)
+		if err != nil {
+			return fmt.Errorf("failed to estimate tx weight for "+
+				"batch %d: %w", batch.id, err)
+		}
+		batches = append(batches, *batchEntity)
+	}
+
+	// Run the algorithm. Get batchId of the best batch, or specialId if the
+	// best option is to create new batch.
+	batchId, err := selectBatch(batches, *sweepEntity, *newBatchEntity)
+	if err != nil {
+		return fmt.Errorf("batch selection algorithm failed for sweep "+
+			"%x: %w", sweep.swapHash[:6], err)
+	}
+
+	// If the best option is to start new batch, do it.
+	if batchId == specialId {
+		return b.launchNewBatch(ctx, sweep)
+	}
+
+	// Locate the batch to add the sweep to.
+	batch, has := b.batches[batchId]
+	if !has {
+		return fmt.Errorf("batch selection algorithm returned "+
+			"batch id %d which doesn't exist, for sweep %x",
+			batchId, sweep.swapHash[:6])
+	}
+
+	// Add the sweep to the batch.
+	accepted, err := batch.addSweep(ctx, sweep)
+	if err != nil {
+		return fmt.Errorf("batch selection algorithm returned "+
+			"batch id %d for sweep %x, but adding failed: %w",
+			batchId, sweep.swapHash[:6], err)
+	}
+	if !accepted {
+		return fmt.Errorf("batch selection algorithm returned "+
+			"batch id %d for sweep %x, but acceptance failed",
+			batchId, sweep.swapHash[:6])
+	}
+
+	return nil
+}
+
+// estimateSweepWeight returns sweep's input weight and one sweep batch weight.
+func estimateSweepWeight(s *sweep) (*entity, *entity, error) {
```

Comment:
nit: perhaps naming this functions as `estimateSweepFeeDelta` or `estimateSweepFeeIncrement` could be more readable. Maybe the godoc can also be made clearer slightly.

Reply:
Renamed to `estimateSweepFeeIncrement` and added more godoc:

```go
// estimateSweepFeeIncrement returns fee details for adding the sweep to a batch
// and for creating new batch with this sweep only.
func estimateSweepFeeIncrement(s *sweep) (feeDetails, feeDetails, error) {
```